include mssql_clustered dialect_options when reflecting - issue #8288

johnlx opened the pull request:
SQL Server: include index clustering info when retrieving via get_indexes()

<!-- Provide a general summary of your proposed changes in the Title field above -->

### Description
<!-- Describe your changes in detail -->
Fixes: #8288 

Ran into this issue when retrieving the list of indexes from one SQL Server database with the `inspect` module and attempting to use that info to build the same onto a second database whose schema had drifted (differences between dev and prod systems).
 
The return values do not include the clustering info for the index which prevents the ability to build this same index on a second database.

More details can be seen in the accompanying issue but the current output looks like this: (no info re: clustering)
```
[{'name': 'x1_testtbl', 'unique': False, 'column_names': ['col1'], 'include_columns': [], 'dialect_options': {'mssql_include': []}}]
```

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

[ ] similar condition may exist with primary keys and should be checked

-->

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

johnlx committed:
Fixes Issue #8288

SQL Server: include index clustering info when retrieving via get_indexes()

johnlx committed:
Fixes Issue #8288

SQL Server: include index clustering info when retrieving via get_indexes()

Fixes issue #8288

SQL Server: include mssql_clustered dialect_options when reflecting
primary key constraint via get_pk_constraint()

Fixes issue #8288

include mssql_clustered value when reflecting
SQL Server index and primary key.

Fixes issue #8288

Formatting changes

Change-Id: I3afe99876fc4c9485bad5ef0bfeb4037a1d6ecf3

zzzeek commented:
any chance you can implement unique constraint and primary key constraint as well?

johnlx commented:
@zzzeek  - sure, I'll take a look at the unique constraints and PKs.
I meant to add a comment that this probably would be required for the primary keys as well anyways, but never considered the unique constraints.

zzzeek commented:
thanks!  I'm just basing this on looking at the original feature and seeing that it applies to those three types of constraints.

johnlx committed:
Fixes issue #8288

SQL Server: include mssql_clustered dialect_options when reflecting
primary key constraint via get_pk_constraint()

johnlx commented:
@zzzeek 
I have some changes implemented for get_pk_constraint() so that the dialect_options will be included in the event that the PK is nonclustered.
e.g.: {'constrained_columns': ['id', 'c1'], 'name': 'pk_testtbl', 'dialect_options': {'mssql_clustered': False} }

_maybe this should be discussed in the accompanying ticket instead of here - let me know...)_  
I do have one question about the approach I took, however, and was curious about your thoughts. 
When reflecting an index, the mssql_clustered option is only included when the index is **clustered**.
Conversely, when reflecting a PK, the mssql_clustered option is only included when the PK is **nonclustered**.

I'm now thinking this feels a little inconsistent, but the reasoning for this was to try to remain inline with the existing behavior of Index and PrimaryKeyConstraint that appear to allow SQL Server's default behavior (Index=nonclustered, PK=clustered) to control whether to cluster the object unless explicitly passed in. 

see this comment in dialects/mssql/base.py

`.. versionchanged:: 1.1 the ``mssql_clustered`` option now defaults to None, rather than False.  ``mssql_clustered=False`` now explicitly renders the NONCLUSTERED clause, whereas None omits the CLUSTERED clause entirely, allowing SQL Server defaults to take effect.`

It's likely that for reflection, though, this could be seen as confusing and maybe it's better to explicitly include True|False for the mssql_clustered option in all cases.

If the reflected data is used to produce a DDL statement, this may have the side-effect of producing SQL like:
`create NONCLUSTERED index x_t on t (Id)`  - where previously the NONCLUSTERED keyword would have been omitted. 


There are a few existing unit-tests that would need to be updated, but it might be preferable to provide more clarity/consistency since we're retrieving the properties of an existing object. The side-effect mentioned above may be a non-issue in practice, especially if it's uncommon to use the reflected data to create the object.

johnlx commented:
Also, with regards to the unique constraints, it doesn't look like the reflection of those was ever implemented - or at least, I am getting a NotImplementedError() when calling: `inspect(engine).get_unique_constraints('testtbl')
`

Not sure if maybe that was an intentional omission since every unique constraint will also have an accompanying unique index, which is already returned by `get_indexes()`. There are some subtle differences between the two, so maybe a separate Issue # would be preferable to see if `get_unique_constraints()` needs to be implemented?

zzzeek commented:
OK, for mssql_nonclustered, I think we should include it in all cases.

for get_unique_constraints(), we're off the hook (translation, no action to be taken now).   some databases combine unique constraints / indexes together as the same thing so I guess MSSQL is one of them.

zzzeek reviewed (changes requested):
you did a lot of work on this thanks very much!  very in-depth tests.     so we just get the mssql_clustered to come back and we can move this to the next review stage.

johnlx commented:
sounds good - thanks for reviewing my questions. I'll work on getting those changes made. I'm hoping they will simplify my earlier changes a bit since some of the checks can be removed.

johnlx committed:
Fixes issue #8288

include mssql_clustered value when reflecting
SQL Server index and primary key.

johnlx commented on test/dialect/mssql/test_reflection.py:
```diff
@@ -566,12 +566,157 @@ def test_indexes_with_filtered(self, metadata, connection):
+        t2 = Table("t", MetaData(), autoload_with=connection)
+        idx = list(sorted(t2.indexes, key=lambda idx: idx.name))[0]
+        self.assert_compile(
+            CreateIndex(idx), "CREATE NONCLUSTERED INDEX idx_x ON t (x) WHERE ([x]='test')"
```
this was an existing test that needed to be updated to include the NONCLUSTERED keyword in the DDL statement since the reflected index now contains that info

johnlx commented on test/dialect/mssql/test_reflection.py:
```diff
@@ -566,12 +566,157 @@ def test_indexes_with_filtered(self, metadata, connection):
 
         self.assert_compile(
-            CreateIndex(idx), "CREATE INDEX idx_x ON t (x) WHERE ([x]='test')"
+            CreateIndex(idx), "CREATE NONCLUSTERED INDEX idx_x ON t (x)"
```
this was an existing test that needed to be updated to include the NONCLUSTERED keyword in the DDL statement since the reflected index now contains that info

johnlx committed:
Fixes issue #8288

Formatting changes

johnlx commented on test/dialect/mssql/test_reflection.py:
```diff
@@ -568,10 +568,166 @@ def test_indexes_with_filtered(self, metadata, connection):
+        )
 
         self.assert_compile(
-            CreateIndex(idx), "CREATE INDEX idx_x ON t (x) WHERE ([x]='test')"
```
this was an existing test that needed to be updated to include the NONCLUSTERED keyword in the DDL statement since the reflected index now contains that info

johnlx commented on test/dialect/mssql/test_reflection.py:
```diff
@@ -568,10 +568,166 @@ def test_indexes_with_filtered(self, metadata, connection):
         idx = list(sorted(t2.indexes, key=lambda idx: idx.name))[0]
+        self.assert_compile(
+            CreateIndex(idx),
+            "CREATE NONCLUSTERED INDEX idx_x ON t (x) WHERE ([x]='test')",
```
this was an existing test that needed to be updated to include the NONCLUSTERED keyword in the DDL statement since the reflected index now contains that info

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision f78fe15c89b72bb5f1df856564625535de51800f of this pull request into gerrit so we can run tests and reviews and stuff

zzzeek commented:
oh great, all our gerrit integration is broken

CaselIT commented:
Strange, I created a gerrit earlier today https://github.com/sqlalchemy/sqlalchemy/pull/8180 and it worked...

zzzeek commented:
OH.  it's the dependabot
 

we cant merge those from here, sorry

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision f78fe15c89b72bb5f1df856564625535de51800f of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision f78fe15c89b72bb5f1df856564625535de51800f of this pull request into gerrit so we can run tests and reviews and stuff

CaselIT commented:
Not sure how it could have broken the gerrit thing.
I had to oass on gerrit since I could not push in that pr branch

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision f78fe15c89b72bb5f1df856564625535de51800f of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision eda62440750f270266a87a1f6d81f04709cc6a0b of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision eda62440750f270266a87a1f6d81f04709cc6a0b of this pull request into gerrit so we can run tests and reviews and stuff

zzzeek commented:
OK the prtogerrit is still broken even though I force pushed a rebase.

it's up at  https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4014 for the moment

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision eda62440750f270266a87a1f6d81f04709cc6a0b of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision eda62440750f270266a87a1f6d81f04709cc6a0b of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset eda62440750f270266a87a1f6d81f04709cc6a0b added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4015

zzzeek commented:
> Not sure how it could have broken the gerrit thing. I had to oass on gerrit since I could not push in that pr branch

if you push a PR to github, it isn't on the gerrit repo, it has to be pushed there separately otherwise nothing will work

CaselIT commented:
I thought I had pushed it to gerrit. Not sure what happened

johnlx committed:
Fixes Issue #8288

Correcting Flake8 formatting issues

johnlx committed:
Merge branch 'feature-mssql-clustered' of https://github.com/johnlx/sqlalchemy into feature-mssql-clustered

johnlx commented:
Hello, I noticed there were still some formatting tests failing. I think I may not have had pre-commit installed correctly.
In any event, I have corrected those errors and the pre-commit tests are passing. Just pushed those to my local branch.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision e62dd8479e8f5bb5dae5732650b635c0abb5f88f of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset e62dd8479e8f5bb5dae5732650b635c0abb5f88f added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4015

johnlx commented:
Hi @zzzeek,
Just thought I would check in to see if there was anything else you need from me on this.
I see that there appears to be a ci_verification issue but I couldn't find the details of the problem, so wasn't sure if I still have some outstanding issues to resolve.

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4015

zzzeek commented:
OK well I cant see why the test failed because the results get deleted after a few days, I just re-requested the tests run again and we will see why they are failing.

zzzeek commented:
> Hi @zzzeek, Just thought I would check in to see if there was anything else you need from me on this. I see that there appears to be a ci_verification issue but I couldn't find the details of the problem, so wasn't sure if I still have some outstanding issues to resolve.

the test that is failing is:

test.dialect.test_suite.ComponentReflectionTestExtra_mssql+pyodbc_14_0_3356_20.test_reflect_covering_index

the output is:

```
AssertionError: [{'name': 't_idx', 'unique': False, 'column_names': ['x'], 'include_columns': ['y'], 'dialect_options': {'mssql_clustered': False, 'mssql_include': ['y']}}] != [{'name': 't_idx', 'column_names': ['x'], 'include_columns': ['y'], 'unique': False, 'dialect_options': {'mssql_include': ['y']}}]
assert [{'column_nam...'t_idx', ...}] == [{'column_nam...'t_idx', ...}]
  At index 0 diff: {'name': 't_idx', 'unique': False, 'column_names': ['x'], 'include_columns': ['y'], 'dialect_options': {'mssql_clustered': False, 'mssql_include': ['y']}} != {'name': 't_idx', 'column_names': ['x'], 'include_columns': ['y'], 'unique': False, 'dialect_options': {'mssql_include': ['y']}}
  Full diff:
    [
     {'column_names': ['x'],
  -   'dialect_options': {'mssql_include': ['y']},
  +   'dialect_options': {'mssql_clustered': False, 'mssql_include': ['y']},
  ?                       ++++++++++++++++++++++++++...
  
  ...Full output truncated (5 lines hidden), use '-vv' to show
```

now, that is not a mis-operation of the new feature you've added, that's a dialect-agnostic test that has expectations that are too narrow for the form of data returned to it.   so here, that test would need adjustment to not compare the "dialect_options" dictionary so rigidly.    here's one way that could be done expediently (though it would be nice if mock had a dict-like structure we could use):

```diff
diff --git a/lib/sqlalchemy/testing/suite/test_reflection.py b/lib/sqlalchemy/testing/suite/test_reflection.py
index a3737a91a1..7e54ee57a0 100644
--- a/lib/sqlalchemy/testing/suite/test_reflection.py
+++ b/lib/sqlalchemy/testing/suite/test_reflection.py
@@ -2396,20 +2396,25 @@ class ComponentReflectionTestExtra(ComparesIndexes, fixtures.TestBase):
 
         insp = inspect(connection)
 
+        get_indexes = insp.get_indexes("t")
         eq_(
-            insp.get_indexes("t"),
+            get_indexes,
             [
                 {
                     "name": "t_idx",
                     "column_names": ["x"],
                     "include_columns": ["y"],
                     "unique": False,
-                    "dialect_options": {
-                        "%s_include" % connection.engine.name: ["y"]
-                    },
+                    "dialect_options": mock.ANY,
                 }
             ],
         )
+        eq_(
+            get_indexes[0]["dialect_options"][
+                "%s_include" % connection.engine.name
+            ],
+            ["y"],
+        )
 
         t2 = Table("t", MetaData(), autoload_with=connection)
         eq_(
```

feel free to add the above patch and we can re-run, if you could test locally that would be helpful.

johnlx committed:
Merge branch 'sqlalchemy:main' into feature-mssql-clustered

johnlx committed:
Fixes Issue #8288

Added patch from @zzzeek to address reflection test

johnlx commented:
Thanks @zzzeek. 

That test issue was a miss on my part. I don't think I ran the dialect\test_suite.py.  I believe I had just been running the mssql specific dialect tests.

Anyways, the failing tests are passing locally now, so hopefully it will get past the CI process this time.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 1bb57352e3e31d8fb7de69ab5e60e5464949f640 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 1bb57352e3e31d8fb7de69ab5e60e5464949f640 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4015

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

I made a few small refinements if you want to check

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4015

johnlx commented:
> **mike bayer** (zzzeek) wrote:
> I made a few small refinements if you want to check

Looks good. Thanks again for the assistance with getting this change included - and your willingness to add it :)

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4015 has been **merged**. Congratulations! :)
